### PR TITLE
[rllib] [experimental] custom RL training pipelines (PG_pl, A2C_pl)

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This adds basic operators for a distributed training DSL based on Ray Parallel Iterators: https://ray.readthedocs.io/en/latest/iter.html

Both PG and A2C are ported. There are some things that don't work such as timestep reporting, but this is a first iteration for feedback.

cc @eugenevinitsky @richardliaw @sven1977

### CODE REVIEW

File: rllib/agents/pg/pg_workflow.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+"""Experimental workflow-based impl; run this with --run='PG_wf'"""
+
+from ray.rllib.agents.pg.pg import PGTrainer
+from ray.rllib.utils.experimental_dsl import (
+    ParallelRollouts, ConcatBatches, TrainOneStep, StandardMetricsReporting)
+
+
+def training_workflow(workers, config):
```
Comment (richardliaw) [inline rllib/agents/pg/pg_workflow.py]: training_workflow => dataflow? or flow?

File: rllib/agents/a3c/a2c_workflow.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    if config["microbatch_size"]:
+        num_microbatches = math.ceil(
+            config["train_batch_size"] / config["microbatch_size"])
+
+        train_op = rollouts \
+            .combine(ConcatBatches(min_batch_size=config["microbatch_size"])) \
+            .for_each(ComputeGradients(workers)) \
+            .batch(num_microbatches) \
+            .for_each(AverageGradients()) \
+            .for_each(ApplyGradients(workers))
```
Comment (richardliaw) [inline rllib/agents/a3c/a2c_workflow.py]: it's hard for me to understand what "each" refers to - if it is referring to batch, what about "for batch"? "for each batch"?

Comment (ericl) [reply rllib/agents/a3c/a2c_workflow.py]: How about .then()?

This is coming from the parallel iterators API, so it is not specific to RL terms (i.e., should not call it for_batch).

File: rllib/agents/a3c/a2c_workflow.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    else:
+        train_op = rollouts \
+            .combine(ConcatBatches(
+                min_batch_size=config["train_batch_size"])) \
+            .for_each(TrainOneStep(workers))
```
Comment (richardliaw) [inline rllib/agents/a3c/a2c_workflow.py]: It's hard to know what operators are operating on "Batches", "Gradients", and "Workers".

**Review by sven1977 — APPROVED** [review]

File: rllib/agents/a3c/a2c_workflow.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+def training_workflow(workers, config):
+    rollouts = ParallelRollouts(workers, mode="bulk_sync")
+
+    if config["microbatch_size"]:
+        num_microbatches = math.ceil(
+            config["train_batch_size"] / config["microbatch_size"])
+
+        train_op = (
+            rollouts.combine(
```
[the remark this answers is not in the document]
Comment (ericl) [reply rllib/agents/a3c/a2c_workflow.py]: Done. Added it in more detail for PG since that is the example algo.

**Comment by ericl** [comment]
Will make new PR for followups.

**Pull request merged**